<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Introduction to rCRUX</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to rCRUX</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rCRUX)</span></code></pre></div>
<p>rCRUX is a suite of tools for querying NCBI databases and filtering
and formatting the resulting data. The typical workflow involves using
get_seeds_local() or get_seeds_remote() to simulate <em>in silico</em>
PCR to acquire a set of sequences analogous to PCR products containing
metabarcode primer sequences. The sequences or “seeds” recovered from
the <em>in silico</em> PCR step are used to search databases for
complementary sequence that lack one or both primers. This search step,
blast_seeds() is used to iteratively align seed sequences against a
local NCBI database for matches using a taxonomic rank based stratified
random sampling approach. This step results in a comprehensive database
of primer specific reference barcode sequences from NCBI. Using
derep_and_clean_db(), the database is de-replicated by DNA sequence
where identical sequences are collapsed into a representative read. If
there are multiple possible taxonomic paths for a read, the taxonomic
path is collapsed to the lowest taxonomic agreement.</p>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>RCRUX can be installed using the following R code:</p>
<pre class="devtools"><code>
devtools::install_github(&quot;LunaGal/rCRUX&quot;)
</code></pre>
<div id="dependencies" class="section level2">
<h2>Dependencies</h2>
<p><strong>NOTE:</strong> These only need to be downloaded once or as
NCBI updates databases. rCRUX can access and successfully build
metabarcode references using databases stored on external drives.
</br></p>
<div id="blast" class="section level3">
<h3>BLAST+</h3>
<p>NCBI’s <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/">BLAST+</a>
suite must be locally installed and accessible in the user’s path. NCBI
provides installation instructions for <a href="https://www.ncbi.nlm.nih.gov/books/NBK52637/">Windows</a>, <a href="https://www.ncbi.nlm.nih.gov/books/NBK52640/">Linux</a>, and <a href="https://www.ncbi.nlm.nih.gov/books/NBK569861/">Mac OS</a>. Version
2.10.1+ is verified compatible with rCRUX.</p>
</div>
<div id="blast-formatted-database" class="section level3">
<h3>Blast-formatted database</h3>
<p>rCRUX requires a local blast-formatted nucleotide database. These can
be user generated or download a pre-formatted database from <a href="https://ftp.ncbi.nlm.nih.gov/blast/db/">NCBI</a>. NCBI provides a
tool (perl script) for downloading databases as part of the blast+
package. A brief help page can be found <a href="https://www.ncbi.nlm.nih.gov/books/NBK569850/">here</a>.</p>
<div id="possible-error-include-but-are-not-limited-to" class="section level4">
<h4>Possible error include but are not limited to:</h4>
<ol style="list-style-type: decimal">
<li><p>Partial downloads of database files. Extracting each TAR archive
(e.g. nt.00.tar.gz.md5) should result in 8 files with the following
extensions(.nhd, .nhi, .nhr, .nin, .nnd, .nni, .nog, and .nsq). If a few
archives fail during download, you can re-download and unpack only those
that failed. You do not have to re-download all archives.</p></li>
<li><p>You downloaded and built a blast database from ncbi fasta files
but did not specify -parse_seqids</p></li>
</ol>
<p>The nt database is <strong>~242 GB</strong> (as of 8/31/22) and can
take several hours (overnight) to build. Loss of internet connection can
lead to partially downloaded files and blastn errors (see above).</p>
</div>
</div>
<div id="taxonomizr" class="section level3">
<h3>Taxonomizr</h3>
<p>rCRUX uses the <a href="https://cran.r-project.org/web/packages/taxonomizr/vignettes/usage.html">taxonomizr</a>
package for taxonomic assignment based on NCBI <a href="https://www.ncbi.nlm.nih.gov/taxonomy">Taxonomy id’s <span class="math inline">\(taxids\)</span></a>. Many rCRUX functions require
a path to a local taxonomizr readable sqlite database. This database can
be built using taxonomizr’s <a href="https://www.rdocumentation.org/packages/taxonomizr/versions/0.8.0/topics/prepareDatabase">prepareDatabase</a>
function. This database is <strong>~72 GB</strong> (as of 8/31/22) and
can take several hours (overnight) to build. Loss of internet connection
can lead to partially downloaded files and taxonomizr run errors.</p>
<p>The following code can be used to build this database:</p>
<pre class="set"><code>
library(taxonomizr)

accession_taxa_sql_path &lt;- &quot;/my/accessionTaxa.sql&quot;
prepareDatabase(accession_taxa_sql_path)
</code></pre>
<div id="note" class="section level4">
<h4>Note:</h4>
<p>For poor bandwidth connections, please see the <a href="https://cran.r-project.org/web/packages/taxonomizr/readme/README.html">taxononmizr
readme for manual installation</a> of the accessionTaxa.sql database. If
built manually, make sure to delete any files other than the
accessionTaxa.sql database (e.g. keeping nucl_gb.accession2taxid.gz
leads to a warning message).</p>
</div>
</div>
</div>
</div>
<div id="example-pipeline" class="section level1">
<h1>Example pipeline</h1>
<p>The following example shows a simple rCRUX pipeline from start to
finish. Note that this example will require internet access and
considerable database storage (~<strong>314 GB</strong>, see section
above), run time (mainly for blastn), and system resources to
execute.</p>
<p>##Note: Blast databases and the taxonomic assignment databases
(accessionTaxa.sql) can be stored on external hard drive. It increases
run time, but is a good option if computer storage capacity is
limited.</p>
<p>There are two options to generate seeds for the database generating
blast step blast_seeds_local() or blast_seeds_remote(). The local option
is slower, however it is not subject to the memory limitations of using
the NCBI primer_blast API. The local option is recommended if the user
is building a large database, wants to include any <a href="https://www.ncbi.nlm.nih.gov/taxonomy">taxid</a> in the search,
and has many degenerate sites in their primer set.</p>
<div id="get_seeds_local" class="section level2">
<h2>get_seeds_local()</h2>
<p>This example uses default parameters, with the exception of evalue to
minimize run time.</p>
<pre class="run"><code>
forward_primer_seq = &quot;TAGAACAGGCTCCTCTAG&quot;
reverse_primer_seq =  &quot;TTAGATACCCCACTATGC&quot;
output_directory_path &lt;- &quot;/my/directory/12S_V5F1_local_111122_e300&quot; # path to desired output directory
metabarcode_name &lt;- &quot;12S_V5F1&quot; # desired name of metabarcode locus
accession_taxa_sql_path &lt;- &quot;/my/directory/accessionTaxa.sql&quot; # path to taxonomizr sql database
blast_db_path &lt;- &quot;/my/directory/ncbi_nt/nt&quot;  # path to blast formatted database


get_seeds_local(forward_primer_seq,
     reverse_primer_seq,
     output_directory_path,
     metabarcode_name,
     accession_taxa_sql_path,
     blast_db_path, evalue = 300)
</code></pre>
<div id="note-1" class="section level4">
<h4>Note:</h4>
<p>Two output .csv files are automatically created at this path based on
the arguments passed to get_seeds_local. One includes all unfiltered
output the other is filtered based on user defined parameters and
includes taxonomy.</p>
<p>A unique taxonomic rank summary file is also generated (e.g. the
number of unique phyla, class, etc in the blast hits). If a taxonomic
rank category contains NA’s, they will be counted as a single unique
rank. Sequence availability in NCBI for a given taxid is a limiting
factor.</p>
<p>Example output can be found <a href="https://github.com/LunaGal/rCRUX/tree/main/examples/12S_V5F1_generated_11-11-22">here</a>.</p>
</div>
<div id="if-blast-is-not-in-your-path-do-the-following" class="section level4">
<h4>If BLAST+ is not in your path do the following:</h4>
<pre class="add"><code>
get_seeds_local(forward_primer_seq,
                 reverse_primer_seq,
                 output_directory_path,
                 metabarcode_name,
                 accession_taxa_sql_path,
                 blast_db_path, evalue = 300,
                 ncbi_bin = &quot;/my/directory/blast+_folder/bin&quot;)
</code></pre>
</div>
</div>
<div id="get_seeds_remote" class="section level2">
<h2>get_seeds_remote()</h2>
<p>This example uses default parameters to minimize run time.</p>
<p>Searching jawless vertebrates (taxid: “1476529”) and jawed
vertebrates (taxid: “7776”).</p>
<pre class="run"><code>
forward_primer_seq = &quot;TAGAACAGGCTCCTCTAG&quot;

reverse_primer_seq =  &quot;TTAGATACCCCACTATGC&quot;

output_directory_path &lt;- &quot;/my/directory/12S_V5F1_remote_111122&quot; # path to desired output directory

metabarcode_name &lt;- &quot;12S_V5F1&quot; # desired name of metabarcode locus

accession_taxa_sql_path &lt;- &quot;/my/directory/accessionTaxa.sql&quot; # path to taxonomizr sql database



get_seeds_remote(forward_primer_seq,
          reverse_primer_seq,
          output_directory_path,
          metabarcode_name,
          accession_taxa_sql_path,
          organism = c(&quot;1476529&quot;, &quot;7776&quot;),
          return_table = FALSE)
</code></pre>
<div id="note-2" class="section level4">
<h4>Note:</h4>
<p>When using default parameters only 1047 hits are returned from NCBI’s
primer blast (run 11-11-22). Returns hit sizes and contents are variable
depending on parameters, random blast sampling, and database
updates.</p>
<p>Two output .csv files are automatically created at this path based on
the arguments passed to get_seeds_remote. One includes all unfiltered
output the other is filtered based on user defined parameters and
includes taxonomy.</p>
<p>A unique taxonomic rank summary file is also generated (e.g. the
number of unique superkingdon, phyla, class, etc in the blast hits). If
a taxonomic rank category contains NA’s, they will be counted as a
single unique rank.</p>
<p>Sequence availability in NCBI for a given taxid is a limiting factor,
as are degenerate bases and API memory allocation.</p>
<p><strong>Modifying defaults</strong> can increase the number of
returns by orders of magnitude.</p>
<p>Example output can be found <a href="https://github.com/LunaGal/rCRUX/tree/main/examples/12S_V5F1_generated_11-11-22">here</a>.</p>
</div>
</div>
<div id="blast_seeds" class="section level2">
<h2>blast_seeds()</h2>
<p>Iterative searches are based on a stratified random sampling unique
taxonomic groups for a given rank from the get_seeds_local or
get_seeds_remote output table. For example, the default is to randomly
sample one read from each genus. The user can select any taxonomic rank
present in the get_seeds_local output table. The number of seeds
selected may cause blastn to exceed the users available RAM, and for
that reason the user can choose the maximum number of reads to blast at
one time (max_to_blast, default = 1000). blast_seeds will subsample each
set of seeds based on max_to_blast and process all seeds before starting
a new search for seeds to blast. It saves the output from each round of
blastn.</p>
<pre class="run"><code>
seeds_output_path &lt;- &#39;/my/directory/12S_V5F1_remote_111122/12S_V5F1_filtered_get_seeds_remote_output_with_taxonomy.csv&#39; # this is output from get_seeds_local or get_seeds_remote

blast_db_path &lt;- &quot;/my/directory/blast_database/nt&quot;  # path to blast formatted database

accession_taxa_sql_path &lt;- &quot;/my/directory/accessionTaxa.sql&quot;  # path to taxonomizr sql database

output_directory_path &lt;- &#39;/my/directory/12S_V5F1_remote_111122/&#39; # path to desired output directory

metabarcode_name &lt;- &quot;12S_V5F1&quot;  # desired name of metabarcode locus


blast_seeds(seeds_output_path,
            blast_db_path,
            accession_taxa_sql_path,
            output_directory_path,
            metabarcode_name)    
</code></pre>
<div id="note-3" class="section level4">
<h4>Note:</h4>
<p>After each round of blast, the system state is saved. If the script
is terminated after a full round of blast, the user can pick up where
they left off. The user can also change parameters at this point
(e.g. change the max_to_blast or rank)</p>
<p>The output includes a summary table of all unique blast hits
(summary.csv), a multi fasta file of all unique hits
(metabarcode_name_.fasta), a taxonomy file of all unique hits
(metabarcode_name_taxonomy.txt), a unique taxonomic rank summary file
(metabarcode_name_taxonomic_rank_counts.txt), a list of all of the
accessions not present in your blast database (e.g. relevant if you ran
get_seeds_remote; blastdbcmd_failed.csv), and a list of accessions with
4 or more Ns in a row (default for that parameter is wildcards = “NNNN”;
too_many_ns.csv). The default number of reads to blast per rank is 1
(default for that parameter is sample_size = 1). The script will error
out if the user asks for more reads per rank than exist in the blast
seeds table.</p>
<p><strong>If BLAST+ is not in your path do the following</strong></p>
<pre class="specify"><code>
blast_seeds(seeds_output_path,
            blast_db_path,
            accession_taxa_sql_path,
            output_directory_path,
            metabarcode_name,
            ncbi_bin = &quot;/my/directory/blast+_folder/bin&quot;)
</code></pre>
</div>
<div id="note-4" class="section level4">
<h4>Note:</h4>
<p>Example output can be found <a href="https://github.com/LunaGal/rCRUX/tree/main/examples/12S_V5F1_generated_11-11-22">here</a>.</p>
<p>There will be variability between runs due to primer blast return
parameters and random sampling of the blast seeds table that occurs
during blast_seeds. However, variability can be decreased by changing
parameters (e.g. randomly sampling species rather than genus will
decrease run to run variability).</p>
</div>
</div>
<div id="derep_and_clean_db" class="section level2">
<h2>derep_and_clean_db()</h2>
<p>This function takes the output of blast_seeds and de-replicates
identical sequences and collapses ambiguous taxonomy to generate a clean
reference database.</p>
<pre class="run"><code>
output_directory_path &lt;- &#39;/my/directory/12S_V5F1_remote_111122/&#39; # path to desired output directory

summary_path &lt;- &quot;/my/directory/12S_V5F1_remote_111122/blast_seeds_output/summary.csv&quot; # this is the path to the output from blast_seeds

derep_and_clean_db(output_directory_path, summary_path, metabarcode_name)
</code></pre>
<div id="note-5" class="section level4">
<h4>Note:</h4>
<p>Accessions with the same sequence are collapsed into a representative
sequence. If those accessions have different taxids (taxonomic paths),
we determine the lowest taxonomic agreement across the multiple
accessions with an identical sequence. For example, for the MiFish 12S
locus, nearly all rockfishes in the genus <em>Sebastes</em> have
identical sequences. Instead of including ~110 identical reference
sequences, one for each individual species, we report a single
representative sequence with a lowest common taxonomic agreement of the
genus <em>Sebastes</em>. This prevents classification bias for taxa with
more sequences and also provides accurate taxonomic resolution within
the reference database.</p>
<p>We exclude all sequences with taxids that are NA. Such sequences are
not immediately useful for classification of metabarcoding sequences.
However, we caution that such results can be indicative of off target
amplification of a given primer set. For example, the MiFish 12S primer
set amplifies uncultured marine bacteria among other taxa (taxid = NA)
indicating off target amplification of non-fish taxa. These sequences
are saved in the References_with_NA_for_taxonomic_ranks.csv file.</p>
<p>The result of this function is a final clean reference database file
set composed of a paired metabarcode_name_derep_and_clean.fasta and
metabarcode_name_derep_and_clean_taxonomy.txt. A summary file of the
number of unique taxonomic ranks is also generated:
metabarcode_name_derep_and_clean_unique_taxonomic_rank_counts.txt. In
addition, all representative sequences and associated accessions are
saved in Sequences_with_lowest_common_taxonomic_path_agreement.csv,
Sequences_with_mostly_NA_taxonomic_paths.csv,
Sequences_with_multiple_taxonomic_paths.csv, and
Sequences_with_single_taxonomic_path.csv files. These files allow for
the traceback of representative sequences to multiple accessions.</p>
<p>Example output can be found <a href="https://github.com/LunaGal/rCRUX/tree/main/examples/12S_V5F1_generated_11-11-22">here</a>.</p>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
