<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Introduction to RCRUX</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to RCRUX</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RCRUX.dev)</span></code></pre></div>
<p>RCRUX is a suite of tools for querying NCBI databases and filtering
and formatting the resulting data. The typical data pipeline involves
using [RCRUX.dev::get_blast_seeds()] to download and wrangle results
from NCBI’s primer BLAST tool, then using [RCRUX.dev::rcrux_blast()] to
search a local NCBI database for matches.</p>
<div id="example-pipeline" class="section level1">
<h1>Example pipeline</h1>
<p>The following example shows a simple RCRUX pipeline from start to
finish. Note that this example would take a lot of storage, time, and
system resources to execute.</p>
<pre><code>seeds_path &lt;- &quot;/my/directory&quot;
accession_path &lt;- &quot;/my/accession.sql&quot;
blastdb_path &lt;- &quot;/my/local/database/nt&quot;

get_blast_seeds(&quot;TAGAACAGGCTCCTCTAG&quot;, &quot;TTAGATACCCCACTATGC&quot;,
                 blast_seeds_parent, &quot;12S_V5F1&quot;, accession_path,
                 organism = c(&quot;7776&quot;, &quot;7777&quot;), return_table = FALSE)

# A .csv is automatically created at this path based on the arguments
# passed to get_blast_seeds
csv_path &lt;- &quot;/my/directory/12S_V5F1/12S_V5F1_primerTree_output_with_taxonomy.csv&quot;

rcrux_blast(csv_path, &quot;blast_test_save&quot;, db_path, accession_taxa_path)</code></pre>
<p>This script takes a forward and reverse primer sequence and generates
a csv with data about matches found in NCBI’s nt database for further
analysis.</p>
</div>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>RCRUX can be installed using the following R code:</p>
<pre><code>devtools::install_github(&quot;LunaGal/RCRUX.dev&quot;)</code></pre>
<p>RCRUX requires several local files in addition to the package
itself.</p>
<div id="blast" class="section level2">
<h2>BLAST+</h2>
<p>NCBI’s BLAST+ suite must be locally installed. NCBI provides
installation instructions for <a href="https://www.ncbi.nlm.nih.gov/books/NBK52637/">Windows</a>, <a href="https://www.ncbi.nlm.nih.gov/books/NBK52640/">Linux</a>, and <a href="https://www.ncbi.nlm.nih.gov/books/NBK569861/">Mac OS</a>.
Additionally, it is available in the Debian archives as ncbi-blast+ and
as a homebrew formula called blast.</p>
</div>
<div id="blast-formatted-database" class="section level2">
<h2>Blast-formatted database</h2>
<p>RCRUX requires a local blast-formatted database. NCBI provides a tool
for downloading databases as part of the blast+ package. A brief help
page can be found <a href="https://www.ncbi.nlm.nih.gov/books/NBK569850/">here</a>.</p>
</div>
<div id="taxonomizr" class="section level2">
<h2>Taxonomizr</h2>
<p>RCRUX uses the taxonomizr package. taxonomizr requires a local sqlite
database and provides prepareDatabase to automatically build the local
database. Many RCRUX functions require a path to a taxonomizr database,
so you should run prepareDatabase before running RCRUX functions. You
can read about taxonomizr <a href="https://cran.r-project.org/web/packages/taxonomizr/vignettes/usage.html">here</a>.</p>
</div>
</div>
<div id="get_blast_seeds" class="section level1">
<h1><a href="https://lunagal.github.io/get_blast_seeds">get_blast_seeds</a></h1>
<p>get_blast_seeds uses modified versions of functions from the
primerTree package to submit queries to NCBI’s primer BLAST tool, then
aggregates results into a single data.frame. primer_search expands
degenerate primers into each possible non-degenerate primer and submits
a query for each. get_blast_seeds further multiplies the number of
queries by allowing the user to query the primers for each organism in a
vector. get_blast_seeds collects all these results from primer_search,
filters them based on product length, and adds taxonomic data using the
taxonomizr package.</p>
<div id="search-options" class="section level2">
<h2>Search options</h2>
<p>get_blast_seeds passes many parameters to NCBI’s primer blast tool.
You can match the parameters to the fields available in the GUI here.
First, use your browser to view the page source. Search for the field
you are interested in by searching for the title of the field. It should
be enclosed in a tag. Inside the label tag, it says for =
“<name_of_parameter>”. Copy the string after for = and add it to
get_blast_seeds as the name of a parameter, setting it equal to whatever
you like.</p>
<p>As of 2022-08-16, the primer blast GUI contains some options that are
not implemented by primer_search. primer_search doesn’t include explicit
documentation of allowed options, but it will quickly report if an
option isn’t allowed, so trial and error will not be very time
consuming.</p>
<p>Example: I want to set “Exon junction span” to 10. I open the source
of the primer designing tool and look for that string. I find the
following:</p>
<p><code>&lt;label class=&quot;m&quot; for=&quot;PRIMER_ON_SPLICE_SITE&quot;&gt;Exon junction span&lt;/label&gt;</code></p>
<p>I copy PRIMER_ON_SPLICE_SITE and add it to get_blast_seeds:</p>
<pre><code>get_blast_seeds(&quot;TAGAACAGGCTCCTCTAG&quot;, &quot;TTAGATACCCCACTATGC&quot;,
             blast_seeds_parent, &quot;12S_V5F1&quot;, accession_taxa_path,
             organism = c(&quot;7776&quot;), MAX_TARGET_PER_TEMPLATE = 10,
             PRIMER_ON_SPLICE_SITE = &quot;10&quot;
             return_table = FALSE)</code></pre>
<div id="organism" class="section level3">
<h3>Organism</h3>
<p>primer BLAST defaults to homo sapiens, so it is important that you
supply a specific organism or organisms. NCBI’s taxids can be found <a href="https://www.ncbi.nlm.nih.gov/taxonomy">here</a>. You can specify
multiple organism by passing a character vector containing each of the
options, like in the example below.</p>
</div>
</div>
<div id="example" class="section level2">
<h2>Example</h2>
<p>This example runs two queries, one for each organism, and writes a
.csv that combines the results.</p>
<pre><code># These file directories need to be changed to locations on your device
blast_seeds_parent &lt;- &quot;D:/blast_seeds_test&quot;
accession_taxa_path &lt;- &quot;D:/taxonomizr_data/accessionTaxa.sql&quot;

get_blast_seeds(&quot;TAGAACAGGCTCCTCTAG&quot;, &quot;TTAGATACCCCACTATGC&quot;,
                blast_seeds_parent, &quot;12S_V5F1&quot;, accession_taxa_path,
                organism = c(&quot;7776&quot;, &quot;7777&quot;), minimum_length = 60,\
                MAX_TARGET_SIZE = 125, return_table = FALSE)</code></pre>
</div>
</div>
<div id="rcrux_blast" class="section level1">
<h1><a href="https://lunagal.github.io/rcrux_blast">rcrux_blast</a></h1>
<p>rcrux_blast uses the entries generated by get_blast_seeds and the
nucleotide-nucleotide matching of blastn to generate a .csv of ncbi
database entries that match a sequence found in the get_blast_seeds
step.</p>
<div id="internal-data-pipeline" class="section level2">
<h2>Internal data pipeline</h2>
<p>rcrux_blast is a wrapper function that passes data to
blast_datatable. rcrux_blast handles the creation of a hidden save
directory and an output directory and writes a .csv summarizing the
results of blast_datatable to the output directory. Optionally, it also
writes .csvs detailing entries rejected by blast_datatable.</p>
<p>Internally, blast_datatable repeatedly samples rows from the table of
seeds, calls blastdbcmd on each accession number, and uses blastn to
build a table of nucleotide matches. It samples by drawing random
indices from a list of unsampled indices and examining the rows at those
indices. It passes those rows to run_blastdbcmd, which extracts the
accession number, forward stop, and reverse stop, then uses them as
arguments for blastdbcmd. blastdbcmd outputs a fasta, which
blast_datatable aggregates into a multi-fasta character vector.
blast_datatable purges any entry that has more than a specified number
of Ns or did not return a result, recording those indices. When it has
finished building the mutli-fasta vector, it passes it to blastn, which
returns every nucleotide sequence that matches a sequence in the file.
run_blastn parses the blastn output into a data.frame, and
blast_datatable adds that data.frame to its output. It repeats this
process until it has sampled every row. Then, it uses taxonomizr to add
taxonomic data to the data.frame based on the accession numbers. The
final output is the aggregate of all blastn calls with the taxonomic
data added.</p>
</div>
<div id="example-1" class="section level2">
<h2>Example</h2>
<p>In this example, rcrux_blast is called on the .csv generated by the
get_blast_seeds example above. This example does not rely on an internet
connection, but it will use a lot of memory and CPU time.</p>
<pre><code># These file directories need to be changed to locations on your device
blast_seeds_parent &lt;- &quot;D:/blast_seeds_test&quot;
accession_taxa_path &lt;- &quot;D:/taxonomizr_data/accessionTaxa.sql&quot;

# This path is indepedent of device; it only depends on get_blast_seeds
# having been run with Metabarcode_name = &quot;12S_V5F1&quot;
seeds_csv_path &lt;- paste0(blast_seeds_parent, &quot;/12S_V5F1/12S_V5F1_raw_primerTree_output.csv&quot;)
RCRUX.dev::rcrux_blast(&quot;short_test/12S_V5F1/12S_V5F1_primerTree_output_with_taxonomy.csv&quot;,
                      &quot;blast_test_save&quot;, db_path, accession_taxa_path)</code></pre>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
