---
title: "Introduction to RCRUX"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to RCRUX}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(RCRUX.dev)
```

RCRUX is a suite of tools for querying NCBI databases and filtering and formatting the resulting data. The typical data pipeline involves using [RCRUX.dev::get_blast_seeds()] to download and wrangle results from NCBI's primer BLAST tool, then using [RCRUX.dev::rcrux_blast()] to search a local NCBI database for matches.

# Example pipeline

The following example shows a simple RCRUX pipeline from start to finish. Note that this example would take a lot of storage, time, and system resources to execute.

```
seeds_path <- "/my/directory"
accession_path <- "/my/accession.sql"
blastdb_path <- "/my/local/database/nt"

get_blast_seeds("TAGAACAGGCTCCTCTAG", "TTAGATACCCCACTATGC",
                 blast_seeds_parent, "12S_V5F1", accession_path,
                 organism = c("7776", "7777"), return_table = FALSE)

# A .csv is automatically created at this path based on the arguments
# passed to get_blast_seeds
csv_path <- "/my/directory/12S_V5F1/12S_V5F1_primerTree_output_with_taxonomy.csv"

rcrux_blast(csv_path, "blast_test_save", db_path, accession_taxa_path)
```

This script takes a forward and reverse primer sequence and generates a csv with data about matches found in NCBI's nt database for further analysis.

# Installation

RCRUX can be installed using the following R code:

```
devtools::install_github("LunaGal/RCRUX.dev")
```

RCRUX requires several local files in addition to the package itself.

## BLAST+

NCBI's BLAST+ suite must be locally installed. NCBI provides installation instructions for [Windows](https://www.ncbi.nlm.nih.gov/books/NBK52637/), [Linux](https://www.ncbi.nlm.nih.gov/books/NBK52640/), and [Mac OS](https://www.ncbi.nlm.nih.gov/books/NBK569861/). Additionally, it is available in the Debian archives as ncbi-blast+ and as a homebrew formula called blast.

## Blast-formatted database

RCRUX requires a local blast-formatted database. NCBI provides a tool for downloading databases as part of the blast+ package. A brief help page can be found [here](https://www.ncbi.nlm.nih.gov/books/NBK569850/).

## Taxonomizr

RCRUX uses the taxonomizr package. taxonomizr requires a local sqlite database and provides prepareDatabase to automatically build the local database. Many RCRUX functions require a path to a taxonomizr database, so you should run prepareDatabase before running RCRUX functions. You can read about taxonomizr [here](https://cran.r-project.org/web/packages/taxonomizr/vignettes/usage.html).

# get_blast_seeds

get_blast_seeds uses modified versions of functions from the primerTree package to submit queries to NCBI's primer BLAST tool, then aggregates results into a single data.frame. primer_search expands degenerate primers into each possible non-degenerate primer and submits a query for each. get_blast_seeds further multiplies the number of queries by allowing the user to query the primers for each organism in a vector. get_blast_seeds collects all these results from primer_search, filters them based on product length, and adds taxonomic data using the taxonomizr package.

## Example

This example runs two queries, one for each organism, and writes a .csv that combines the results.

```
# These file directories need to be changed to locations on your device
blast_seeds_parent <- "D:/blast_seeds_test"
accession_taxa_path <- "D:/taxonomizr_data/accessionTaxa.sql"

get_blast_seeds("TAGAACAGGCTCCTCTAG", "TTAGATACCCCACTATGC",
                blast_seeds_parent, "12S_V5F1", accession_taxa_path,
                organism = c("7776", "7777"), return_table = FALSE)
```

# rcrux_blast

rcrux_blast uses the entries generated by get_blast_seeds and the nucleotide-nucleotide matching of blastn to generate a .csv of ncbi database entries that match a sequence found in the get_blast_seeds step.

## Internal data pipeline

rcrux_blast is a wrapper function that passes data to blast_datatable. rcrux_blast handles the creation of a hidden save directory and an output directory and writes a .csv summarizing the results of blast_datatable to the output directory. Optionally, it also writes .csvs detailing entries rejected by blast_datatable.

Internally, blast_datatable repeatedly samples rows from the table of seeds, calls blastdbcmd on each accession number, and uses blastn to build a table of nucleotide matches. It samples by drawing random indices from a list of unsampled indices and examining the rows at those indices. It passes those rows to run_blastdbcmd, which extracts the accession number, forward stop, and reverse stop, then uses them as arguments for blastdbcmd. blastdbcmd outputs a fasta, which blast_datatable aggregates into a multi-fasta character vector. blast_datatable purges any entry that has more than a specified number of Ns or did not return a result, recording those indices. When it has finished building the mutli-fasta vector, it passes it to blastn, which returns every nucleotide sequence that matches a sequence in the file. run_blastn parses the blastn output into a data.frame, and blast_datatable adds that data.frame to its output. It repeats this process until it has sampled every row. Then, it uses taxonomizr to add taxonomic data to the data.frame based on the accession numbers. The final output is the aggregate of all blastn calls with the taxonomic data added.

## Example

In this example, rcrux_blast is called on the .csv generated by the get_blast_seeds example above. This example does not rely on an internet connection, but it will use a lot of memory and CPU time.

```
# These file directories need to be changed to locations on your device
blast_seeds_parent <- "D:/blast_seeds_test"
accession_taxa_path <- "D:/taxonomizr_data/accessionTaxa.sql"

# This path is indepedent of device; it only depends on get_blast_seeds
# having been run with Metabarcode_name = "12S_V5F1"
seeds_csv_path <- paste0(blast_seeds_parent, "/12S_V5F1/12S_V5F1_raw_primerTree_output.csv")
RCRUX.dev::rcrux_blast("short_test/12S_V5F1/12S_V5F1_primerTree_output_with_taxonomy.csv",
                      "blast_test_save", db_path, accession_taxa_path)
```