<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Runs blastn with the input primer sequences converted to a...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">

<table width="100%" summary="page for run_primer_blastn {tools}"><tr><td>run_primer_blastn {tools}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Runs blastn with the input primer sequences converted to a primer_fasta file
as a query</h2>

<h3>Description</h3>

<p>Runs blastn with the input primer sequences converted to a primer_fasta file
as a query
</p>


<h3>Usage</h3>

<pre>
run_primer_blastn(
  primer_fasta,
  db,
  ncbi_bin = NULL,
  task = "blastn-short",
  word_size = 7,
  evalue = "3e+07",
  coverage = 90,
  perID = 50,
  reward = 2
)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>primer_fasta</code></td>
<td>
<p>path to the primer fasta file</p>
</td></tr>
<tr valign="top"><td><code>db</code></td>
<td>
<p>path to blast formatted database</p>
</td></tr>
<tr valign="top"><td><code>ncbi_bin</code></td>
<td>
<p>is the path to blast+ tools if not in the user's path.
Specify only if blastn and blastdbcmd  are not in your path.
The default is ncbi_bin = NULL - if not specified in path do the
following: ncbi_bin = &quot;/my/local/blast+_folder&quot;.</p>
</td></tr>
<tr valign="top"><td><code>task</code></td>
<td>
<p>the task for blastn to perform - default here is &quot;blastn_short&quot;,
which is optimized for searches with queries &lt; 50 bp</p>
</td></tr>
<tr valign="top"><td><code>word_size</code></td>
<td>
<p>is the fragment size used for blastn search - smaller word
sizes increase sensitivity and time of the search.
The default is word_size =  7</p>
</td></tr>
<tr valign="top"><td><code>evalue</code></td>
<td>
<p>is the number of expected hits with a similar quality score
found by chance. The default is evalue = 3e-7.</p>
</td></tr>
<tr valign="top"><td><code>coverage</code></td>
<td>
<p>is the minimum percent of the query length recovered in the
subject hits. The default is coverage = 90.</p>
</td></tr>
<tr valign="top"><td><code>perID</code></td>
<td>
<p>is the minimum percent identity of the query relative to the
subject hits. The default is perID = 50.</p>
</td></tr>
<tr valign="top"><td><code>reward</code></td>
<td>
<p>is the reward for nucleotide match. The default is reward = 2.</p>
</td></tr>
</table>


<h3>Details</h3>

<p>Calls blastn with a primer fasta file as the query. The user can not add
additional search parameters, but can modify the available parameters.
</p>
<p>run_primer_blastn takes a fasta file containing primers, uses blastn-short to
query them to a blast formatted database. The result is an output table with
the following columns of data: qseqid (query subject id), sgi (subject gi),
saccver (subject accession version), mismatch (number of mismatches between
the subject a query), sstart (subject start), send (subject end), staxids
(subject taxids).
</p>
<p>Information about the blastn parameters can be found by accessing blastn -help
and at <a href="https://www.ncbi.nlm.nih.gov/books/NBK279684/">NCBI</a>.
</p>
<p>Note:
The number of alignments returned for a given blast search is hardcoded at
&quot;-num_alignments 10000000&quot;.
</p>


<h3>Value</h3>

<p>a tibble 'output_table' representing the blastn results
</p>

<hr /><div style="text-align: center;">[Package <em>tools</em> version 4.1.2 ]</div>
</div></body></html>
